k2/csrc/cuda/debug_test.cu

```diff
@@ -0,0 +1,190 @@
+// k2/csrc/cuda/debug_test.cu
+
+// Copyright (c) 2020, Xiaomi Corporation ( authors: Meixu Song )
+
+// See ../../LICENSE for clarification regarding multiple authors
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#ifndef K2_PARANOID
+#define K2_PARANOID
+#endif
+
+#include "k2/csrc/cuda/debug.h"
+
+namespace k2 {
+
+// A dummy cuda asynchronous function
+__global__ void FillContents(int n, int *output) {
+  int correct_index = static_cast<int>(threadIdx.x) * n;
+
+  for (int i = correct_index; i < n; i++) {
+    output[i] = (i - correct_index) * 2;
+    K2_DLOG("Print from async cuda kernel, output[%d] = %d\n", i, output[i]);
+  }
+}
+
+/* number of blocks (max 65,536) */
+#define BLOCKS 50000
+
+/* number of threads per block (max 1,024) */
+#define THREADS 200
+
+/* N is now the number of blocks times the number of threads
+   in this case it is 10 million */
+#define N (BLOCKS * THREADS)
+
+/* the kernel fills an array up with square roots */
+__global__ void square_roots(double* array) {
+  /* find my array index
+     this now uses three values:
+     blockIdx.x which is the block ID we are on
+     blockDim.x which is the number of blocks in the X dimension
+     threadIdx.x which is the thread ID we are */
+  unsigned int idx = blockIdx.x * blockDim.x + threadIdx.x;
+
+  /* compute the square root of this number and store it in the array
+     CUDA provides a sqrt function as part of its math library, so this
+     call uses that - not the one from math.h which is a CPU function */
+  array[idx] = sqrt((double) idx);
+}
+
+__global__ void HelloCUDA(float f) {
+  if (threadIdx.x == 0)
+    K2_DLOG("Hello thread %d, f=%f\n", threadIdx.x, f);
+}
+
+// A vector add kernel definition
+__global__ void VecAdd(float* A, float* B, float* C) {
```

Comment:
Please add `const` to function arguments that you do not intend to modify inside the function.

Reply:
This function and its argument init are from <<CUDA C Programming Guide>> Chapter 2. I would like to untouch it.